### ISSUE TITLE
Move Debian packaging in separate branch

### ISSUE THREAD

**Issue by Tachi107** [pistacheio/pistache#954]
Instead of moving the Debian changes in a separate branch, we could move them to a separate repo, like `pistache-debian`.

This would allow us to use `git-buildpackage`, a tool that automates the job of keeping the two repos in sync. @kiplingw, you can take a brief look at it in the latest [Guide for Debian Maintainers](https://www.debian.org/doc/manuals/debmake-doc/), specifically in [chapter 7](https://www.debian.org/doc/manuals/debmake-doc/ch07.en.html).

I've already set up a test environment on my own GitHub profile; [Tachi107/pistache](https://github.com/Tachi107/pistache) represents the upstream repo, while [Tachi107/pistache-debian](https://github.com/Tachi107/pistache-debian) represents the Debian packaging one. The latter consists in three branches, `master` containing the source + debian changes, `upstream` containing only the untouched upstream source, and, optionally, `pristine-tar` containing the history of all releases.

This way packaging would be done the same way it is done on [salsa.debian.org](https://salsa.debian.org), the official place where actual packaging happens ([example](https://salsa.debian.org/debian/curl)).

I think this would be one of the few solutions that would allow us to keep building for the PPA, while removing the Debian changes from the repo thus making downstream packaging easier and more realistic.

I already tried moving the `debian/` dir to a separate branch, but keeping it in sync with the main one is a huge pain; `git-buildpackage` is simply awesome.

**Comment by Tachi107** [comment]
Maybe the only issue is that we would need organization-level privileges to create a new repo, and it seems that the organization has no (public) members, so as far as I can tell @oktal is the only one who could create it.

Mathieu, could you please add (at least) one of the maintainers to the organization? (if this is already the case, sorry for the useless ping)

**Comment by dennisjenkins75** [comment]
IMHO, distribution-specific packaging files should not be in the main project anyway.  I rarely see OSS projects with specific rpm spec files; and Gentoo places all of their ebuilds into their own repo.

**Comment by kiplingw** [comment]
A lot of free software projects do store packaging metadata within their repositories. There are literally hundreds of them. A common one I use daily is Nicotine+. This is partially because of the prevalence of PPAs, but also because distros sometimes take months or even years to release new stable versions, even if upstream's cycle might be as little as a few weeks at times.

I'm generally in agreement with @dennisjenkins75. Where I disagree is that packaging metadata needs to be provided by upstream (somehow or somewhere) until the software makes its way downstream where someone else at some point takes over package maintenance. That hasn't happened yet. After we have our first stable release, that's more likely to happen. Even Debian is generally only again storing the metadata in the repository if it's already accepted downstream. Before that time they don't really have a policy.

**Comment by Tachi107** [comment]
> dennisjenkins75, can you or oktal please create a new empty repository called pistache-debian? Tachi107 can then submit two PRs, one to remove from the main branch and another for the new one.

I think it would be better to directly push the changes without merge commits or similar things (just in case `git-buildpackage` expects a somehow clean history); I can create the git repo on my machine, send the archive to one of you with write permissions and then you can do something like `git remote add origin git@github.com:pistacheio/pistache-debian.git`, `git push --all --set-upstream origin`

If Dennis can't create new repos in the organization and Mathieu doesn't respond in the next few days, as a temporary solution we could create the packaging repo in one of our private profiles (@kiplingw's would be the most appropriate), since it is not necessary that they are on the same profile. Not clean, but better than nothing

**Comment by kiplingw** [comment]
@Tachi107, I didn't realize my bits were already set to create the new repository. Check your email. You should be good to go.

@pabs3, do you have any suggestions for his suggested workflow above? We have an unstable and stable PPA, latter not having anything in it yet. We'd like to be able to continue having binary packages produced in the PPA, but we also don't want to keep the Debianization or any other packaging metadata in the main repository.

**Comment by Tachi107** [comment]
@kiplingw it seems that I can't push to the repo, don't know if this is intentional. If it is, tomorrow I'll send you an archive containing the stuff that you have to push

**Comment by kiplingw** [comment]
Try now. Just set your admin bit on `debian-pistache`.

**Comment by Tachi107** [comment]
I'm reading launchpad docs, as soon as I'm ready I'll start doing stuff

**Comment by Tachi107** [comment]
Ok, to move the packaging data to another repo without breaking anything we need to:

- Delete `debian/` from [pistacheio/pistache](https://github.com/pistacheio/pistache)
- Tag a new release
- Download a tarball of the release and debianize it locally
- Generate a .dsc file from the debianized release
- Run `gbp import-dsc --pristine-tar pistache*.dsc` in a new directory to generate a git repository with the correct branches (master, upstream, pristine-tar)
- Push the repo with all its branches to [pistacheio/pistache-debian](https://github.com/pistacheio/pistache-debian) (`git remote add origin git@github.com:pistacheio/pistache-debian.git`, `git push --all --set-upstream origin`)
- Update the launchpad import repo ([here](https://code.launchpad.net/~pistache+team/pistache/+git/main/+edit-import))

This would only work for the stable PPA, as `git-buildpackage` (or rather, `uscan`) uses debian/watch to check for updates and thus would only update pistacheio/pistache-debian on every new tag. Maybe a custom recipe could help with the unstable PPA

**Comment by kiplingw** [comment]
Sounds reasonable. The tagging of a new release could actually be good timing since we were planning on doing that any ways.

I'm going to leave this in your hands @Tachi107, as this was your idea. But I think you've thought it through well enough that I will merge your PR when ready. 

The unstable PPA probably will need its own custom recipe, but that has to be done before we break anything because my test beds, servers, and workstations are all dependant on the unstable PPA right now.

**Comment by pabs3** [comment]
I'm of the opinion that *sometimes* it is useful to keep the `debian/` dir in the upstream repo; when you want to be able to build packages from any commit in the repo during CI. If you split it out, it becomes more tricky to package anything other than the latest release. Alphas/betas and "release early release often" solves that issue though.

That said the plan looks reasonable to me.

**Comment by kiplingw** [comment]
The Downstream Debian Deity hath spoken.

**Comment by Tachi107** [comment]
Hi everyone, I'm back :)

After looking a bit into how Launchpad works I think that my original plan is a bit more complicated than it needs to be. In fact, simply moving the Debian packaging data to another branch works perfectly fine for daily builds ([recipe here](https://code.launchpad.net/~tachi107/+recipe/pistache-tachi-test-daily#edit-recipe_text)). 

I don't know how to automatically handle stable builds, but setting the stable ppa to never build automatically and start a build only when a new release is made wouldn't be all that bad. @kiplingw do you have any better idea?

**Comment by kiplingw** [comment]
Hey @Tachi107. Welcome back. I think the easiest thing to do is to let the unstable PPA build as it normally does. When it's time to release a stable, you can literally copy the generated packages from the unstable to stable PPA. There's a UI for it in Launchpad. This might not be elegant if we had a hundred releases we were putting out every year, but I can't imagine us updating the stable PPA very often (consider that this is our first time).

**Comment by Tachi107** [comment]
Alright, now the only thing preventing us from moving the Debian data is the autopkgtest GitHub Action. As soon as I update that code to download Debian data from the `debian` branch I'll be ready to prepare the PR

**Comment by Tachi107** [comment]
Hi @kiplingw, since I don't have write permissions I can't create a clean `debian` branch in this repo. Could you please create it for me?

```sh
git checkout --orphan debian
git rm --force $(git ls-files)
git commit --allow-empty --message='Create debian branch'
git push --set-upstream origin debian
```

**Comment by Tachi107** [comment]
Oh and could you also delete some old branches? They have been already merged in master, and you can find them in [pistacheio/pistache/branches/stale](https://github.com/pistacheio/pistache/branches/stale).

`headers`
`promise-refactor`
`io-refactor`
`dennisjenkins75-patch-readme-contact-info`
`revert-236-master`
`travis-enforce-unit-tests`
`dennisjenkins75-warning-redux`
`dennisjenkins75-cleanup-1`
`revert-665-kip-fix-pkgconfig-cpp14`
`revert-814-reduce_memory_usage_post_files`
`revert-836-add_peer_id`
`server-client-request-timeout`
`revert-890-meson-improvements-library-pkgconf`

There are other branches that haven't been touched for 6 years, but they haven't been merged in master

**Comment by kiplingw** [comment]
> Hi @kiplingw, since I don't have write permissions I can't create a clean `debian` branch in this repo. Could you please create it for me?
> 
> ```shell
> git checkout --orphan debian
> git rm --force $(git ls-files)
> git commit --allow-empty --message='Create debian branch'
> git push --set-upstream origin debian
> ```

Done.

**Comment by kiplingw** [comment]
> Oh and could you also delete some old branches? They have been already merged in master, and you can find them in [pistacheio/pistache/branches/stale](https://github.com/pistacheio/pistache/branches/stale).
> 
> `headers`
> `promise-refactor`
> `io-refactor`
> `dennisjenkins75-patch-readme-contact-info`
> `revert-236-master`
> `travis-enforce-unit-tests`
> `dennisjenkins75-warning-redux`
> `dennisjenkins75-cleanup-1`
> `revert-665-kip-fix-pkgconfig-cpp14`
> `revert-814-reduce_memory_usage_post_files`
> `revert-836-add_peer_id`
> `server-client-request-timeout`
> `revert-890-meson-improvements-library-pkgconf`
> 
> There are other branches that haven't been touched for 6 years, but they haven't been merged in master

Done.
